sql: TestCachedNodeSequence failed

sql.TestCachedNodeSequence [failed](https://tanzanite.cluster.engflow.com/invocations/default/7997c18f-6e7a-43c9-89c6-b37de47e69ca?testReportRun=16&testReportShard=5&testReportAttempt=1#targets-Ly9wa2cvc3FsOnNxbF90ZXN0) on master @ [41da9eccc216413c09cc66ad4c6bfa5f92c506e3](https://github.com/cockroachdb/cockroach/commits/41da9eccc216413c09cc66ad4c6bfa5f92c506e3):

Fatal error:

```
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0x11dd870]

```
Stack: 

```
goroutine 256547 [running]:
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).closeWrapper(0xc00364cc88, {0x811f008, 0xc01fada7e0}, {0x5df9800?, 0xbd57620?})
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1215 +0x19d
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn.func1({0x811f008, 0xc01fada7e0}, {0xc0054ebc00?})
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:963 +0x45
panic({0x5df9800?, 0xbd57620?})
	GOROOT/src/runtime/panic.go:770 +0x132
github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError.func1()
	github.com/cockroachdb/cockroach/pkg/sql/colexecerror/error.go:58 +0x43d
panic({0x5df9800?, 0xbd57620?})
	GOROOT/src/runtime/panic.go:770 +0x132
github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError.func1()
	github.com/cockroachdb/cockroach/pkg/sql/colexecerror/error.go:58 +0x43d
panic({0x5df9800?, 0xbd57620?})
	GOROOT/src/runtime/panic.go:770 +0x132
github.com/cockroachdb/cockroach/pkg/sql/sessiondatapb.(*SequenceCacheNode).NextValue(0xc020d175c0, 0x68, 0x1, 0xc0054e83e8)
	github.com/cockroachdb/cockroach/pkg/sql/sessiondatapb/pkg/sql/sessiondatapb/sequence_cache_node.go:65 +0x130
github.com/cockroachdb/cockroach/pkg/sql.(*planner).incrementSequenceUsingCache(0xc00364d300, {0x811f008, 0xc01f8f70e0}, {0x8227480, 0xc00aa7ef08})
	github.com/cockroachdb/cockroach/pkg/sql/sequence.go:220 +0x19d
github.com/cockroachdb/cockroach/pkg/sql.incrementSequenceHelper({0x811f008, 0xc01f8f70e0}, 0xc00364d300, {0x8227480, 0xc00aa7ef08})
	github.com/cockroachdb/cockroach/pkg/sql/sequence.go:125 +0x1ea
github.com/cockroachdb/cockroach/pkg/sql.(*planner).IncrementSequenceByID(0xc00364d300, {0x811f008, 0xc01f8f70e0}, 0xc0181bd81e?)
	github.com/cockroachdb/cockroach/pkg/sql/sequence.go:91 +0x125
github.com/cockroachdb/cockroach/pkg/sql/sem/builtins.init.func491({0x811f008, 0xc01f8f70e0}, 0xc01b0da808, {0xc00d08bbe0?, 0x0?, 0x0?})
	github.com/cockroachdb/cockroach/pkg/sql/sem/builtins/builtins.go:2177 +0xb4
github.com/cockroachdb/cockroach/pkg/sql/sem/eval.(*evaluator).EvalFuncExpr(0xc01b0da808, {0x811f008, 0xc01f8f70e0}, 0xc027f925a0)
	github.com/cockroachdb/cockroach/pkg/sql/sem/eval/expr.go:483 +0x114
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*FuncExpr).Eval(0xc01f8f6c90?, {0x811f008?, 0xc01f8f70e0?}, {0x81c3280?, 0xc01b0da808?})
	github.com/cockroachdb/cockroach/bazel-out/k8-fastbuild/bin/pkg/sql/sem/tree/eval_expr_generated.go:281 +0x31
github.com/cockroachdb/cockroach/pkg/sql/sem/eval.Expr(...)
	github.com/cockroachdb/cockroach/pkg/sql/sem/eval/expr.go:26
github.com/cockroachdb/cockroach/pkg/sql.(*valuesNode).startExec(0xc012a3d260, {{0x811f008, 0xc01f8f70e0}, 0xc01b0da808, 0xc00364d300})
	github.com/cockroachdb/cockroach/pkg/sql/values.go:89 +0x2a7
github.com/cockroachdb/cockroach/pkg/sql.startExec.func2({0x4764b8?, 0xc0054e8aa8?}, {0x811c2c0, 0xc012a3d260})
	github.com/cockroachdb/cockroach/pkg/sql/plan.go:520 +0x13d
github.com/cockroachdb/cockroach/pkg/sql.(*planVisitor).visitInternal.func1()
	github.com/cockroachdb/cockroach/pkg/sql/walk.go:112 +0x38
github.com/cockroachdb/cockroach/pkg/sql.(*planVisitor).visitInternal(0xc0054e8b98, {0x811c2c0?, 0xc012a3d260}, {0x67ea9ab?, 0x6?})
	github.com/cockroachdb/cockroach/pkg/sql/walk.go:299 +0x1602
github.com/cockroachdb/cockroach/pkg/sql.(*planVisitor).visit(0xc0054e8b98, {0x811c2c0, 0xc012a3d260})
	github.com/cockroachdb/cockroach/pkg/sql/walk.go:79 +0xf0
github.com/cockroachdb/cockroach/pkg/sql.walkPlan(...)
	github.com/cockroachdb/cockroach/pkg/sql/walk.go:43
github.com/cockroachdb/cockroach/pkg/sql.startExec({{0x811f008?, 0xc01f8f70e0?}, 0xc01b0da808?, 0xc00364d300?}, {0x811c2c0, 0xc012a3d260})
	github.com/cockroachdb/cockroach/pkg/sql/plan.go:523 +0x10b
github.com/cockroachdb/cockroach/pkg/sql.(*planNodeToRowSource).Start(0xc006cc2e08, {0x811f008, 0xc01f8f7080})
	github.com/cockroachdb/cockroach/pkg/sql/plan_node_to_row_source.go:175 +0x10d
github.com/cockroachdb/cockroach/pkg/sql/colexec.(*Columnarizer).Init(0xc008f5e608, {0x811f008, 0xc01f8f7020})
	github.com/cockroachdb/cockroach/pkg/sql/colexec/columnarizer.go:183 +0xe8
github.com/cockroachdb/cockroach/pkg/sql/colexec.(*invariantsChecker).Init(0xc018da18c0, {0x811f008?, 0xc01f8f7020?})
	github.com/cockroachdb/cockroach/pkg/sql/colexec/invariants_checker.go:67 +0x49
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*batchInfoCollector).init(...)
	github.com/cockroachdb/cockroach/pkg/sql/colflow/stats.go:94
github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError(0xc018da1900?)
	github.com/cockroachdb/cockroach/pkg/sql/colexecerror/error.go:92 +0x57
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*batchInfoCollector).Init(0xc005a64600, {0x811f008?, 0xc01f8f7020?})
	github.com/cockroachdb/cockroach/pkg/sql/colflow/stats.go:103 +0x77
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*BatchFlowCoordinator).Run.(*BatchFlowCoordinator).init.func2()
	github.com/cockroachdb/cockroach/pkg/sql/colflow/flow_coordinator.go:245 +0x28
github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError(0x811efd0?)
	github.com/cockroachdb/cockroach/pkg/sql/colexecerror/error.go:92 +0x57
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*BatchFlowCoordinator).init(...)
	github.com/cockroachdb/cockroach/pkg/sql/colflow/flow_coordinator.go:244
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*BatchFlowCoordinator).Run(0xc007192b40, {0x811efd0?, 0xc01d10dd10?})
	github.com/cockroachdb/cockroach/pkg/sql/colflow/flow_coordinator.go:278 +0x108
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*vectorizedFlow).Run(0xc0075b3440, {0x811efd0, 0xc01d10dd10}, 0x0)
	github.com/cockroachdb/cockroach/pkg/sql/colflow/vectorized_flow.go:320 +0x1f3
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run(0xc031230f00, {0x811f008, 0xc01f8f6d80}, 0xc008853380, 0xc006f5dcc0, 0xc01f670000, 0xc0083c1808, 0xc00364d7d0, 0xc0054e9a78)
	github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:911 +0xcc7
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRun(0xc031230f00, {0x811f008, 0xc01f8f6d80}, 0xc00364d7d0, 0xc008853380, 0xc006f5dcc0, {{0x811c2c0, 0xc012a3d260}, 0x0}, 0xc0083c1808, ...)
	github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:1972 +0x215
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRunAll.func3(0x2?, 0x10?, {0x811f008?, 0xc01f8f6d80?}, 0x4764b8?, 0x7f69e0cc82f8?, 0xc00d08bb30?)
	github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:1687 +0xbb
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRunAll(0xc031230f00, {0x811f008, 0xc01f8f6d80}, 0xc00364d7d0, 0xc008853380, 0xc00364d300, 0xc0083c1808, 0x0)
	github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:1690 +0x23b
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithDistSQLEngine(0xc00364cc88, {0x811f008, 0xc01f8f6d80}, 0xc00364d300, 0x3?, {0x7f69f4b6f050?, 0xc00b52f330?}, 0x0, 0xc008853298)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:2443 +0x445
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).dispatchToExecutionEngine(0xc00364cc88, {0x811f008, 0xc01f8f6960}, 0xc00364d300, {0x7f69f4b6f050, 0xc00b52f330})
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1975 +0x1272
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState(0xc00364cc88, {0x811f008, 0xc01f8f6960}, {{0x8141e20, 0xc01d10da90}, {0x0, 0x0, 0x0}, {0xc002a79000, 0x14}, ...}, ...)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1147 +0x454d
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt.func1({0x811f008?, 0xc01f8f6750?})
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:146 +0xb0
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithProfiling(0x811f008?, {0x811f008?, 0xc01f8f6750?}, {0x8141e20?, 0xc01d10da90?}, 0x67eb089?, 0xc00b52ec08?)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:3452 +0x185
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt(0xc00364cc88, {0x811f008, 0xc01f8f6750}, {{0x8141e20, 0xc01d10da90}, {0x0, 0x0, 0x0}, {0xc002a79000, 0x14}, ...}, ...)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:145 +0x5bc
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd.func1(0xc009b5f820, 0xc009b5f8f0, 0xc00364cc88, 0xc009b5f518, 0xc0054eb978, 0xc0054eb968, 0xc0054eb998, 0xc009b5f9b0)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:2244 +0x3f1
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd(0xc00364cc88)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:2249 +0xa56
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run(0xc00364cc88, {0x811f008, 0xc01fada7e0}, 0x81899d0?, 0x0?, 0xc001de0f30?)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:2166 +0x208
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn(0xc010a97400?, {0x811f008?, 0xc01fada7e0?}, {0xc002b4c01e?}, 0x4?, 0xc01fada630?)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:965 +0xda
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).processCommands(0xc00b52ec08, {0x811f008, 0xc01fada7e0}, {0x0, 0x4, {0x1, {0x67d8195, 0x3}, {0xc0174fed70, 0xf}, ...}, ...}, ...)
	github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:247 +0x411
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*Server).serveImpl.func3()
	github.com/cockroachdb/cockroach/pkg/sql/pgwire/server.go:999 +0xec
created by github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*Server).serveImpl in goroutine 256546
	github.com/cockroachdb/cockroach/pkg/sql/pgwire/server.go:996 +0x5f9
```
<details><summary>Log preceding fatal error</summary>
<p>

```
*   | 	github.com/cockroachdb/cockroach/pkg/sql/sessiondatapb/pkg/sql/sessiondatapb/sequence_cache_node.go:65
*   | github.com/cockroachdb/cockroach/pkg/sql.(*planner).incrementSequenceUsingCache
*   | 	github.com/cockroachdb/cockroach/pkg/sql/sequence.go:220
*   | github.com/cockroachdb/cockroach/pkg/sql.incrementSequenceHelper
*   | 	github.com/cockroachdb/cockroach/pkg/sql/sequence.go:125
*   | github.com/cockroachdb/cockroach/pkg/sql.(*planner).IncrementSequenceByID
*   | 	github.com/cockroachdb/cockroach/pkg/sql/sequence.go:91
*   | github.com/cockroachdb/cockroach/pkg/sql/sem/builtins.init.func491
*   | 	github.com/cockroachdb/cockroach/pkg/sql/sem/builtins/builtins.go:2177
*   | github.com/cockroachdb/cockroach/pkg/sql/sem/eval.(*evaluator).EvalFuncExpr
*   | 	github.com/cockroachdb/cockroach/pkg/sql/sem/eval/expr.go:483
*   | github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*FuncExpr).Eval
*   | 	github.com/cockroachdb/cockroach/bazel-out/k8-fastbuild/bin/pkg/sql/sem/tree/eval_expr_generated.go:281
*   | github.com/cockroachdb/cockroach/pkg/sql/sem/eval.Expr
*   | 	github.com/cockroachdb/cockroach/pkg/sql/sem/eval/expr.go:26
*   | github.com/cockroachdb/cockroach/pkg/sql.(*valuesNode).startExec
*   | 	github.com/cockroachdb/cockroach/pkg/sql/values.go:89
*   | github.com/cockroachdb/cockroach/pkg/sql.startExec.func2
*   | 	github.com/cockroachdb/cockroach/pkg/sql/plan.go:520
*   | github.com/cockroachdb/cockroach/pkg/sql.(*planVisitor).visitInternal.func1
*   | 	github.com/cockroachdb/cockroach/pkg/sql/walk.go:112
*   | github.com/cockroachdb/cockroach/pkg/sql.(*planVisitor).visitInternal
*   | 	github.com/cockroachdb/cockroach/pkg/sql/walk.go:299
*   | github.com/cockroachdb/cockroach/pkg/sql.(*planVisitor).visit
*   | 	github.com/cockroachdb/cockroach/pkg/sql/walk.go:79
*   | github.com/cockroachdb/cockroach/pkg/sql.walkPlan
*   | 	github.com/cockroachdb/cockroach/pkg/sql/walk.go:43
*   | github.com/cockroachdb/cockroach/pkg/sql.startExec
*   | 	github.com/cockroachdb/cockroach/pkg/sql/plan.go:523
*   | github.com/cockroachdb/cockroach/pkg/sql.(*planNodeToRowSource).Start
*   | 	github.com/cockroachdb/cockroach/pkg/sql/plan_node_to_row_source.go:175
*   | github.com/cockroachdb/cockroach/pkg/sql/colexec.(*Columnarizer).Init
*   | 	github.com/cockroachdb/cockroach/pkg/sql/colexec/columnarizer.go:183
*   | github.com/cockroachdb/cockroach/pkg/sql/colexec.(*invariantsChecker).Init
*   | 	github.com/cockroachdb/cockroach/pkg/sql/colexec/invariants_checker.go:67
*   | github.com/cockroachdb/cockroach/pkg/sql/colflow.(*batchInfoCollector).init
*   | 	github.com/cockroachdb/cockroach/pkg/sql/colflow/stats.go:94
*   | github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError
*   | 	github.com/cockroachdb/cockroach/pkg/sql/colexecerror/error.go:92
*   | github.com/cockroachdb/cockroach/pkg/sql/colflow.(*batchInfoCollector).Init
*   | 	github.com/cockroachdb/cockroach/pkg/sql/colflow/stats.go:103
*   | github.com/cockroachdb/cockroach/pkg/sql/colflow.(*BatchFlowCoordinator).Run.(*BatchFlowCoordinator).init.func2
*   | 	github.com/cockroachdb/cockroach/pkg/sql/colflow/flow_coordinator.go:245
*   | github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError
*   | 	github.com/cockroachdb/cockroach/pkg/sql/colexecerror/error.go:92
*   | github.com/cockroachdb/cockroach/pkg/sql/colflow.(*BatchFlowCoordinator).init
*   | 	github.com/cockroachdb/cockroach/pkg/sql/colflow/flow_coordinator.go:244
* Wraps: (4) runtime error: invalid memory address or nil pointer dereference
* Error types: (1) *withstack.withStack (2) *safedetails.withSafeDetails (3) *withstack.withStack (4) runtime.errorString
*
```
</p>
</details>

Parameters:
 - <code>attempt=1</code>
 - <code>run=16</code>
 - <code>shard=5</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-queries
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestCachedNodeSequence.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-36521

--- a/pkg/sql/sessiondatapb/sequence_cache_node.go
+++ b/pkg/sql/sessiondatapb/sequence_cache_node.go
@@ -52,7 +52,7 @@ func (sc *SequenceCacheNode) NextValue(
 		sc.mu.Lock()
 		defer sc.mu.Unlock()
 		// There is a hazard that multiple threads could add the entry, so check if it exists again with the writer lock
-		_, found = sc.cache[seqID]
+		cacheEntry, found = sc.cache[seqID]
 		if !found {
 			cacheEntry = &SequenceCacheNodeEntry{}
 			sc.cache[seqID] = cacheEntry

